File: fulfillment-without-a-tracking-number/node/flag-untracked-shipments.js
/**
 * Flag Medusa fulfillments that are shipped but carry no tracking number.
 *
 * In Medusa v2, a fulfillment's tracking data lives in Fulfillment.labels[], each
 * label carrying tracking_number, tracking_url, and label_url, separate from the
 * shipped_at timestamp that actually marks it as shipped. The Admin dashboard's
 * Create Shipment flow, backed by createShipmentWorkflow, has historically built
 * the shipment's labels solely from whatever was typed into that form's tracking
 * number input, discarding any labels a fulfillment provider had already
 * attached in createFulfillment(). Because tracking entry is optional, a
 * merchant can click Mark as Shipped, setting shipped_at, while labels stays
 * empty (see medusajs/medusa issue #11160, partially addressed in PR #11775).
 *
 * There is no legitimate value this script could invent for a missing tracking
 * number, so this only flags and reports. The only write it will ever make is
 * attaching a real label a human has already obtained from the carrier or
 * fulfillment provider, and only when DRY_RUN is off. Run on a schedule. Safe to
 * run again and again.
 *
 * Guide: https://www.allanninal.dev/medusa/fulfillment-without-a-tracking-number/
 */
import { pathToFileURL } from "node:url";

const BACKEND_URL = process.env.MEDUSA_BACKEND_URL || "http://localhost:9000";
const ADMIN_EMAIL = process.env.MEDUSA_ADMIN_EMAIL || "admin@example.com";
const ADMIN_PASSWORD = process.env.MEDUSA_ADMIN_PASSWORD || "supersecret";
const DRY_RUN = (process.env.DRY_RUN || "true").toLowerCase() === "true";

const ORDER_FIELDS = "id,display_id,email,*fulfillments,*fulfillments.labels";

/**
 * Pure decision function. No I/O.
 *
 * A fulfillment is "shipped without tracking" iff:
 *   1. shipped_at is set (truthy) -> it has actually been marked shipped
 *   2. canceled_at is NOT set -> ignore canceled fulfillments (irrelevant once canceled)
 *   3. labels is missing/empty OR every label has a blank tracking_number
 *
 * @param {Array<{ id: string, shipped_at: string | null, canceled_at: string | null,
 *   labels?: Array<{ tracking_number?: string | null }> }>} fulfillments
 * @returns {Array<{ id: string, reason: string }>}
 */
export function findUntrackedShipments(fulfillments) {
  return fulfillments
    .filter((f) => {
      const isShipped = !!f.shipped_at;
      const isCanceled = !!f.canceled_at;
      const hasLabels = Array.isArray(f.labels) && f.labels.length > 0;
      const hasTrackingNumber =
        hasLabels && f.labels.some((l) => !!(l.tracking_number && l.tracking_number.trim().length > 0));
      return isShipped && !isCanceled && !hasTrackingNumber;
    })
    .map((f) => ({ id: f.id, reason: "shipped_at set but no non-empty tracking_number on any label" }));
}

async function getAdminToken() {
  const res = await fetch(`${BACKEND_URL}/auth/user/emailpass`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ email: ADMIN_EMAIL, password: ADMIN_PASSWORD }),
  });
  if (!res.ok) throw new Error(`Medusa auth ${res.status}`);
  const body = await res.json();
  return body.token;
}

async function adminGet(token, path, params = {}) {
  const url = new URL(`${BACKEND_URL}${path}`);
  for (const [key, value] of Object.entries(params)) url.searchParams.set(key, value);
  const res = await fetch(url, { headers: { Authorization: `Bearer ${token}` } });
  if (!res.ok) throw new Error(`Medusa ${res.status}`);
  return res.json();
}

async function adminPost(token, path, body) {
  const res = await fetch(`${BACKEND_URL}${path}`, {
    method: "POST",
    headers: { Authorization: `Bearer ${token}`, "Content-Type": "application/json" },
    body: JSON.stringify(body),
  });
  if (!res.ok) throw new Error(`Medusa ${res.status}`);
  return res.json();
}

async function listOrders(token) {
  const orders = [];
  let offset = 0;
  const limit = 100;
  while (true) {
    const data = await adminGet(token, "/admin/orders", { fields: ORDER_FIELDS, limit, offset });
    orders.push(...data.orders);
    offset += limit;
    if (offset >= data.count) return orders;
  }
}

async function refetchFulfillment(token, orderId, fulfillmentId) {
  // Re-read directly, since list responses can omit nested labels.
  const path = `/admin/orders/${orderId}/fulfillments/${fulfillmentId}`;
  const data = await adminGet(token, path, { fields: "id,shipped_at,canceled_at,*labels" });
  return data.fulfillment;
}

/** The only legitimate corrective write. Never call this with a synthesized value.
 * Mirrors the same route the Admin dashboard's Mark as Shipped / Create Shipment
 * form calls, backed by createShipmentWorkflow.
 */
async function attachTrackingNumber(token, orderId, fulfillmentId, trackingNumber, trackingUrl, labelUrl) {
  const path = `/admin/orders/${orderId}/fulfillments/${fulfillmentId}/shipment`;
  const body = { labels: [{ tracking_number: trackingNumber, tracking_url: trackingUrl, label_url: labelUrl }] };
  return adminPost(token, path, body);
}

function toCsv(rows) {
  const header = "order_id,display_id,fulfillment_id,shipped_at,provider_id";
  const lines = rows.map((r) =>
    [r.order_id, r.display_id, r.fulfillment_id, r.shipped_at, r.provider_id].map((v) => v ?? "").join(",")
  );
  return [header, ...lines].join("\n");
}

export async function run() {
  const token = await getAdminToken();

  const rows = [];
  for (const order of await listOrders(token)) {
    const fulfillments = order.fulfillments || [];
    for (const flagged of findUntrackedShipments(fulfillments)) {
      const fulfillment = fulfillments.find((f) => f.id === flagged.id);
      // Re-read directly before flagging, since list responses can omit nested labels.
      const rechecked = await refetchFulfillment(token, order.id, fulfillment.id);
      if (findUntrackedShipments([rechecked]).length) {
        rows.push({
          order_id: order.id,
          display_id: order.display_id,
          fulfillment_id: fulfillment.id,
          shipped_at: fulfillment.shipped_at,
          provider_id: fulfillment.provider_id,
        });
        console.warn(
          `Order ${order.display_id} fulfillment ${fulfillment.id} shipped with no tracking number. ${DRY_RUN ? "would report" : "reporting"}`
        );
      }
    }
  }

  const report = toCsv(rows);
  console.log(`Done. ${rows.length} fulfillment(s) shipped without tracking.`);
  return report;
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) {
  run().catch((err) => { console.error(err); process.exit(1); });
}

File: fulfillment-without-a-tracking-number/python/flag_untracked_shipments.py
"""Flag Medusa fulfillments that are shipped but carry no tracking number.

In Medusa v2, a fulfillment's tracking data lives in Fulfillment.labels[], each
label carrying tracking_number, tracking_url, and label_url, separate from the
shipped_at timestamp that actually marks it as shipped. The Admin dashboard's
Create Shipment flow, backed by createShipmentWorkflow, has historically built
the shipment's labels solely from whatever was typed into that form's tracking
number input, discarding any labels a fulfillment provider had already
attached in createFulfillment(). Because tracking entry is optional, a
merchant can click Mark as Shipped, setting shipped_at, while labels stays
empty (see medusajs/medusa issue #11160, partially addressed in PR #11775).

There is no legitimate value this script could invent for a missing tracking
number, so this only flags and reports. The only write it will ever make is
attaching a real label a human has already obtained from the carrier or
fulfillment provider, and only when DRY_RUN is off. Run on a schedule. Safe to
run again and again.

Guide: https://www.allanninal.dev/medusa/fulfillment-without-a-tracking-number/
"""
import csv
import io
import os
import logging
import requests

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("flag_untracked_shipments")

BACKEND_URL = os.environ["MEDUSA_BACKEND_URL"]
ADMIN_EMAIL = os.environ["MEDUSA_ADMIN_EMAIL"]
ADMIN_PASSWORD = os.environ["MEDUSA_ADMIN_PASSWORD"]
DRY_RUN = os.environ.get("DRY_RUN", "true").lower() == "true"

ORDER_FIELDS = "id,display_id,email,*fulfillments,*fulfillments.labels"


def get_admin_token():
    r = requests.post(
        f"{BACKEND_URL}/auth/user/emailpass",
        json={"email": ADMIN_EMAIL, "password": ADMIN_PASSWORD},
        timeout=30,
    )
    r.raise_for_status()
    return r.json()["token"]


def admin_get(token, path, params=None):
    r = requests.get(
        f"{BACKEND_URL}{path}",
        headers={"Authorization": f"Bearer {token}"},
        params=params or {},
        timeout=30,
    )
    r.raise_for_status()
    return r.json()


def admin_post(token, path, body):
    r = requests.post(
        f"{BACKEND_URL}{path}",
        headers={"Authorization": f"Bearer {token}", "Content-Type": "application/json"},
        json=body,
        timeout=30,
    )
    r.raise_for_status()
    return r.json()


def find_untracked_shipments(fulfillments):
    """Pure decision function. No I/O.

    fulfillments: list of {id, shipped_at, canceled_at, labels?: [{tracking_number?}]}

    A fulfillment is "shipped without tracking" iff:
      1. shipped_at is set (truthy) -> it has actually been marked shipped
      2. canceled_at is NOT set -> ignore canceled fulfillments (irrelevant once canceled)
      3. labels is missing/empty OR every label has a blank tracking_number

    Returns a list of {id, reason}.
    """
    flagged = []
    for f in fulfillments:
        is_shipped = bool(f.get("shipped_at"))
        is_canceled = bool(f.get("canceled_at"))
        labels = f.get("labels") or []
        has_tracking_number = any(
            (l.get("tracking_number") or "").strip() for l in labels
        )
        if is_shipped and not is_canceled and not has_tracking_number:
            flagged.append({
                "id": f["id"],
                "reason": "shipped_at set but no non-empty tracking_number on any label",
            })
    return flagged


def refetch_fulfillment(token, order_id, fulfillment_id):
    """Re-read a single fulfillment directly, since list responses can omit nested labels."""
    path = f"/admin/orders/{order_id}/fulfillments/{fulfillment_id}"
    data = admin_get(token, path, {"fields": "id,shipped_at,canceled_at,*labels"})
    return data["fulfillment"]


def list_orders(token):
    orders = []
    offset = 0
    limit = 100
    while True:
        data = admin_get(token, "/admin/orders", {
            "fields": ORDER_FIELDS,
            "limit": limit,
            "offset": offset,
        })
        orders.extend(data["orders"])
        offset += limit
        if offset >= data["count"]:
            return orders


def attach_tracking_number(token, order_id, fulfillment_id, tracking_number, tracking_url=None, label_url=None):
    """The only legitimate corrective write. Never call this with a synthesized value.

    Mirrors the same route the Admin dashboard's Mark as Shipped / Create Shipment
    form calls, backed by createShipmentWorkflow.
    """
    path = f"/admin/orders/{order_id}/fulfillments/{fulfillment_id}/shipment"
    body = {"labels": [{
        "tracking_number": tracking_number,
        "tracking_url": tracking_url,
        "label_url": label_url,
    }]}
    return admin_post(token, path, body)


def write_report(rows):
    buf = io.StringIO()
    writer = csv.DictWriter(buf, fieldnames=["order_id", "display_id", "fulfillment_id", "shipped_at", "provider_id"])
    writer.writeheader()
    for row in rows:
        writer.writerow(row)
    return buf.getvalue()


def run():
    token = get_admin_token()

    rows = []
    for order in list_orders(token):
        fulfillments = order.get("fulfillments") or []
        for flagged in find_untracked_shipments(fulfillments):
            fulfillment = next(f for f in fulfillments if f["id"] == flagged["id"])
            # Re-read directly before flagging, since list responses can omit nested labels.
            rechecked = refetch_fulfillment(token, order["id"], fulfillment["id"])
            if find_untracked_shipments([rechecked]):
                rows.append({
                    "order_id": order["id"],
                    "display_id": order.get("display_id"),
                    "fulfillment_id": fulfillment["id"],
                    "shipped_at": fulfillment.get("shipped_at"),
                    "provider_id": fulfillment.get("provider_id"),
                })
                log.warning(
                    "Order %s fulfillment %s shipped with no tracking number. %s",
                    order.get("display_id"), fulfillment["id"],
                    "would report" if DRY_RUN else "reporting",
                )

    report = write_report(rows)
    log.info("Done. %d fulfillment(s) shipped without tracking.", len(rows))
    return report


if __name__ == "__main__":
    run()
